File: app/src/main/java/com/bsoft/inventorymanager/viewmodels/CreateSaleViewModel.java
package com.bsoft.inventorymanager.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.bsoft.inventorymanager.models.Customer;
import com.bsoft.inventorymanager.models.Product;
import com.bsoft.inventorymanager.models.Sale;
import com.bsoft.inventorymanager.models.SaleItem;
import com.bsoft.inventorymanager.repositories.SaleRepository;
import com.bsoft.inventorymanager.utils.FinancialCalculator;
import com.bsoft.inventorymanager.models.ProductSelection;
import com.google.firebase.Timestamp;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class CreateSaleViewModel extends ViewModel {

    private final SaleRepository saleRepository;

    // State
    private final MutableLiveData<List<ProductSelection>> productSelections = new MutableLiveData<>(new ArrayList<>());
    private final MutableLiveData<Double> subtotal = new MutableLiveData<>(0.0);
    private final MutableLiveData<Customer> selectedCustomer = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private final MutableLiveData<String> error = new MutableLiveData<>();
    private final MutableLiveData<String> saleSuccess = new MutableLiveData<>();
    private final MutableLiveData<List<Customer>> customers = new MutableLiveData<>();

    @Inject
    public CreateSaleViewModel(SaleRepository saleRepository) {
        this.saleRepository = saleRepository;
        loadCustomers();
    }

    // Getters
    public LiveData<List<ProductSelection>> getProductSelections() {
        return productSelections;
    }

    public LiveData<List<Customer>> getCustomers() {
        return customers;
    }

    public LiveData<Double> getSubtotal() {
        return subtotal;
    }

    public LiveData<Customer> getSelectedCustomer() {
        return selectedCustomer;
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<String> getError() {
        return error;
    }

    public LiveData<String> getSaleSuccess() {
        return saleSuccess;
    }

    // Actions
    public void setSelectedCustomer(Customer customer) {
        selectedCustomer.setValue(customer);
    }

    public void addProduct(Product product) {
        List<ProductSelection> currentItems = productSelections.getValue();
        if (currentItems == null)
            currentItems = new ArrayList<>();

        // Check if already exists
        boolean found = false;
        for (ProductSelection item : currentItems) {
            if (item.getProduct().getDocumentId().equals(product.getDocumentId())) {
                // Already added - increment
                item.setQuantityInSale(item.getQuantityInSale() + 1);
                found = true;
                break;
            }
        }

        if (!found) {
            ProductSelection newItem = new ProductSelection(product, 1);
            currentItems.add(newItem);
        }

        productSelections.setValue(currentItems);
        recalculateSubtotal();
    }

    public void removeProduct(int position) {
        List<ProductSelection> currentItems = productSelections.getValue();
        if (currentItems != null && position >= 0 && position < currentItems.size()) {
            currentItems.remove(position);
            productSelections.setValue(currentItems);
            recalculateSubtotal();
        }
    }

    public void updateItemQuantity(int position, int quantity) {
        List<ProductSelection> currentItems = productSelections.getValue();
        if (currentItems != null && position >= 0 && position < currentItems.size()) {
            ProductSelection item = currentItems.get(position);
            item.setQuantityInSale(quantity);
            // productSelections.setValue(currentItems); // Prevent loop/focus loss
            recalculateSubtotal();
        }
    }

    public void updateItemPrice(int position, double price) {
        List<ProductSelection> currentItems = productSelections.getValue();
        if (currentItems != null && position >= 0 && position < currentItems.size()) {
            ProductSelection item = currentItems.get(position);
            item.getProduct().setSellingPrice(price);
            // productSelections.setValue(currentItems); // Prevent loop/focus loss
            recalculateSubtotal();
        }
    }

    private void recalculateSubtotal() {
        double sub = 0.0;
        List<ProductSelection> items = productSelections.getValue();
        if (items != null) {
            for (ProductSelection item : items) {
                sub += FinancialCalculator.calculateLineItemTotal(item.getProduct().getSellingPrice(),
                        item.getQuantityInSale());
            }
        }
        subtotal.setValue(sub);
    }

    public void saveSale(Date saleDate, double taxAmount, double discountAmount, String paymentMethod,
            double amountPaid,
            String notes) {
        if (isLoading.getValue() == Boolean.TRUE)
            return;

        List<ProductSelection> selections = productSelections.getValue();
        if (selections == null || selections.isEmpty()) {
            error.setValue("No products selected");
            return;
        }

        if (selectedCustomer.getValue() == null) {
            error.setValue("No customer selected");
            return;
        }

        isLoading.setValue(true);

        Sale sale = new Sale();
        sale.setCustomerId(selectedCustomer.getValue().getDocumentId());
        sale.setCustomerName(selectedCustomer.getValue().getName());

        Date dateParam = saleDate != null ? saleDate : new Date();
        sale.setSaleDate(new Timestamp(dateParam));

        List<SaleItem> items = new ArrayList<>();
        double totalCost = 0.0;
        for (ProductSelection sel : selections) {
            SaleItem item = new SaleItem();
            item.setProductId(sel.getProduct().getDocumentId());
            item.setProductName(sel.getProduct().getName());
            item.setPricePerItem(sel.getProduct().getSellingPrice());
            item.setQuantity(sel.getQuantityInSale());
            item.setCategory(sel.getProduct().getCategory());
            item.setBrand(sel.getProduct().getBrand());

            // Capture historical cost price
            double itemCost = sel.getProduct().getCostPrice();
            item.setCostPrice(itemCost);
            totalCost += (itemCost * sel.getQuantityInSale());

            items.add(item);
        }
        sale.setItems(items);

        double sub = subtotal.getValue() != null ? subtotal.getValue() : 0.0;
        sale.setSubtotal(sub);
        sale.setTaxAmount(taxAmount);
        sale.setDiscountAmount(discountAmount);

        double totalAmount = FinancialCalculator.calculateTotalAmount(sub, taxAmount, discountAmount);
        sale.setTotalAmount(totalAmount);

        // Populate pre-calculated financial data
        sale.setTotalCost(totalCost);
        sale.setTotalProfit((sub - discountAmount) - totalCost);

        sale.setPaymentMethod(paymentMethod);
        sale.setAmountPaid(amountPaid);
        sale.setNotes(notes);
        // Calculate amountDue
        sale.setAmountDue(totalAmount - amountPaid);
        // TODO: Get real user ID
        sale.setUserId("CURRENT_USER_ID");
        sale.setStatus("confirmed"); // Default to confirmed for now

        saleRepository.saveSale(sale, items, new SaleRepository.SaleCallback() {
            @Override
            public void onSuccess(String saleId) {
                isLoading.setValue(false);
                saleSuccess.setValue(saleId);
            }

            @Override
            public void onFailure(Exception e) {
                isLoading.setValue(false);
                error.setValue(e.getMessage());
            }
        });
    }

    public void loadProductByBarcode(String barcode) {
        isLoading.setValue(true);
        saleRepository.getProductByBarcode(barcode, new SaleRepository.ProductCallback() {
            @Override
            public void onSuccess(Product product) {
                isLoading.setValue(false);
                addProduct(product);
            }

            @Override
            public void onFailure(Exception e) {
                isLoading.setValue(false);
                error.setValue("Product not found");
            }
        });
    }

    public void loadSale(String saleId) {
        isLoading.setValue(true);
        saleRepository.getSale(saleId, new SaleRepository.GetSaleCallback() {
            @Override
            public void onSuccess(Sale sale) {
                List<SaleItem> items = sale.getItems();
                if (items == null || items.isEmpty()) {
                    isLoading.setValue(false);
                    return;
                }

                // Reset state
                productSelections.setValue(new ArrayList<>());

                // Customer
                Customer c = new Customer();
                c.setDocumentId(sale.getCustomerId());
                c.setName(sale.getCustomerName());
                // Ideally we should fetch the full customer object here
                selectedCustomer.setValue(c);

                fetchProductsForItems(items, 0, new ArrayList<>());
            }

            @Override
            public void onFailure(Exception e) {
                isLoading.setValue(false);
                error.setValue("Failed to load sale: " + e.getMessage());
            }
        });
    }

    private void loadCustomers() {
        saleRepository.getCustomers(new SaleRepository.CustomersCallback() {
            @Override
            public void onSuccess(List<Customer> customerList) {
                customers.setValue(customerList);
            }

            @Override
            public void onFailure(Exception e) {
                error.setValue("Failed to load customers: " + e.getMessage());
            }
        });
    }

    private void fetchProductsForItems(List<SaleItem> items, int index, List<ProductSelection> accumulated) {
        if (index >= items.size()) {
            productSelections.setValue(accumulated);
            recalculateSubtotal();
            isLoading.setValue(false);
            return;
        }

        SaleItem item = items.get(index);
        saleRepository.getProduct(item.getProductId(), new SaleRepository.ProductCallback() {
            @Override
            public void onSuccess(Product product) {
                // Use fresh product data but maintain sale quantity
                ProductSelection sel = new ProductSelection(product, item.getQuantity());
                // Override selling price with the price from the sale item to reflect
                // historical price
                product.setSellingPrice(item.getPricePerItem());

                accumulated.add(sel);
                fetchProductsForItems(items, index + 1, accumulated);
            }

            @Override
            public void onFailure(Exception e) {
                // If product deleted or not found, create placeholder
                Product placeholder = new Product();
                placeholder.setDocumentId(item.getProductId());
                placeholder.setName(item.getProductName() + " (Unavailable)");
                placeholder.setSellingPrice(item.getPricePerItem());
                placeholder.setQuantity(0);

                ProductSelection sel = new ProductSelection(placeholder, item.getQuantity());
                accumulated.add(sel);
                fetchProductsForItems(items, index + 1, accumulated);
            }
        });
    }
}

File: app/src/main/java/com/bsoft/inventorymanager/viewmodels/CreateSaleViewModel.kt
package com.bsoft.inventorymanager.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bsoft.inventorymanager.model.Customer
import com.bsoft.inventorymanager.models.Product
import com.bsoft.inventorymanager.models.ProductSelection
import com.bsoft.inventorymanager.models.Sale
import com.bsoft.inventorymanager.models.SaleItem
import com.bsoft.inventorymanager.repository.CustomerRepository
import com.bsoft.inventorymanager.repositories.SaleRepository
import com.bsoft.inventorymanager.utils.FinancialCalculator
import com.google.firebase.Timestamp
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import java.util.Date
import javax.inject.Inject

@HiltViewModel
class CreateSaleViewModel @Inject constructor(
    private val saleRepository: SaleRepository,
    private val customerRepository: CustomerRepository
) : ViewModel() {

    // State
    private val _productSelections = MutableLiveData<List<ProductSelection>>(ArrayList())
    val productSelections: LiveData<List<ProductSelection>> = _productSelections

    private val _subtotal = MutableLiveData(0.0)
    val subtotal: LiveData<Double> = _subtotal

    private val _selectedCustomer = MutableLiveData<Customer?>()
    val selectedCustomer: LiveData<Customer?> = _selectedCustomer

    private val _isLoading = MutableLiveData(false)
    val isLoading: LiveData<Boolean> = _isLoading

    private val _error = MutableLiveData<String?>()
    val error: LiveData<String?> = _error

    private val _saleSuccess = MutableLiveData<String?>()
    val saleSuccess: LiveData<String?> = _saleSuccess

    private val _customers = MutableLiveData<List<Customer>>()
    val customers: LiveData<List<Customer>> = _customers

    init {
        loadCustomers()
    }

    // Actions
    fun setSelectedCustomer(customer: Customer?) {
        _selectedCustomer.value = customer
    }

    fun loadCustomerById(customerId: String) {
        _isLoading.value = true
        viewModelScope.launch {
            customerRepository.getCustomer(customerId)
                .onSuccess {
                    _selectedCustomer.value = it
                    _isLoading.value = false
                }
                .onFailure {
                    _error.value = "Failed to load customer: ${it.message}"
                    _isLoading.value = false
                }
        }
    }

    fun addProduct(product: Product) {
        val currentItems = _productSelections.value?.toMutableList() ?: ArrayList()
        
        // Check if already exists
        var found = false
        for (item in currentItems) {
            if (item.product.documentId == product.documentId) {
                // Already added - increment
                item.quantityInSale = item.quantityInSale + 1
                found = true
                break
            }
        }

        if (!found) {
            val newItem = ProductSelection(product, 1)
            currentItems.add(newItem)
        }

        _productSelections.value = currentItems
        recalculateSubtotal()
    }

    fun removeProduct(position: Int) {
        val currentItems = _productSelections.value?.toMutableList()
        if (currentItems != null && position >= 0 && position < currentItems.size) {
            currentItems.removeAt(position)
            _productSelections.value = currentItems
            recalculateSubtotal()
        }
    }

    fun updateItemQuantity(position: Int, quantity: Int) {
        val currentItems = _productSelections.value
        if (currentItems != null && position >= 0 && position < currentItems.size) {
            val item = currentItems[position]
            item.quantityInSale = quantity
            recalculateSubtotal()
        }
    }

    fun updateItemPrice(position: Int, price: Double) {
        val currentItems = _productSelections.value
        if (currentItems != null && position >= 0 && position < currentItems.size) {
            val item = currentItems[position]
            item.product.sellingPrice = price
            recalculateSubtotal()
        }
    }

    private fun recalculateSubtotal() {
        var sub = 0.0
        val items = _productSelections.value
        if (items != null) {
            for (item in items) {
                sub += FinancialCalculator.calculateLineItemTotal(
                    item.product.sellingPrice,
                    item.quantityInSale
                )
            }
        }
        _subtotal.value = sub
    }

    fun saveSale(
        saleDate: Date?, taxAmount: Double, discountAmount: Double, paymentMethod: String,
        amountPaid: Double,
        notes: String
    ) {
        if (_isLoading.value == true) return

        val selections = _productSelections.value
        if (selections == null || selections.isEmpty()) {
            _error.value = "No products selected"
            return
        }

        if (_selectedCustomer.value == null) {
            _error.value = "No customer selected"
            return
        }

        _isLoading.value = true

        val sale = Sale()
        sale.customerId = _selectedCustomer.value!!.documentId
        // Assuming getName() is available on shared Customer model
        sale.customerName = _selectedCustomer.value!!.name

        val dateParam = saleDate ?: Date()
        sale.saleDate = Timestamp(dateParam)

        val items: MutableList<SaleItem> = ArrayList()
        var totalCost = 0.0
        for (sel in selections) {
            val item = SaleItem()
            item.productId = sel.product.documentId
            item.productName = sel.product.name
            item.pricePerItem = sel.product.sellingPrice
            item.quantity = sel.quantityInSale
            item.category = sel.product.category
            item.brand = sel.product.brand

            // Capture historical cost price
            val itemCost = sel.product.costPrice
            item.costPrice = itemCost
            totalCost += (itemCost * sel.quantityInSale)

            items.add(item)
        }
        sale.items = items

        val sub = _subtotal.value ?: 0.0
        sale.subtotal = sub
        sale.taxAmount = taxAmount
        sale.discountAmount = discountAmount

        val totalAmount = FinancialCalculator.calculateTotalAmount(sub, taxAmount, discountAmount)
        sale.totalAmount = totalAmount

        // Populate pre-calculated financial data
        sale.totalCost = totalCost
        sale.totalProfit = (sub - discountAmount) - totalCost

        sale.paymentMethod = paymentMethod
        sale.amountPaid = amountPaid
        sale.notes = notes
        // Calculate amountDue
        sale.amountDue = totalAmount - amountPaid
        // TODO: Get real user ID
        sale.userId = "CURRENT_USER_ID"
        sale.status = "confirmed" // Default to confirmed for now

        saleRepository.saveSale(sale, items, object : SaleRepository.SaleCallback {
            override fun onSuccess(saleId: String) {
                _isLoading.postValue(false)
                _saleSuccess.postValue(saleId)
            }

            override fun onFailure(e: Exception) {
                _isLoading.postValue(false)
                _error.postValue(e.message)
            }
        })
    }

    fun loadProductByBarcode(barcode: String) {
        _isLoading.value = true
        saleRepository.getProductByBarcode(barcode, object : SaleRepository.ProductCallback {
            override fun onSuccess(product: Product) {
                _isLoading.postValue(false)
                // Need to use postValue or runOnMainThread if callback is background. 
                // Assuming SafeRepository callbacks are on main, but using postValue is safer if unsure.
                // However, addProduct accesses LiveData.value which must be main thread.
                // If callback is bg, we crash. 
                // Let's assume callbacks are main thread or use viewModelScope to wrap if needed.
                // For now, I'll assume usage pattern hasn't changed.
               addProduct(product)
            }

            override fun onFailure(e: Exception) {
                _isLoading.postValue(false)
                _error.postValue("Product not found")
            }
        })
    }

    fun loadSale(saleId: String) {
        _isLoading.value = true
        saleRepository.getSale(saleId, object : SaleRepository.GetSaleCallback {
            override fun onSuccess(sale: Sale) {
                val items = sale.items
                if (items == null || items.isEmpty()) {
                    _isLoading.postValue(false)
                    return
                }

                // Reset state
                _productSelections.value = ArrayList()

                // Customer - Need to construct minimal Customer or load it
                // Since CreateSaleActivity uses this to set selectedCustomer, 
                // and we now want KMP customer, we can't easily construct a full mock if KMP model doesn't allow it or if fields are readonly.
                // But KMP data class usually has copy/constructor.
                // However, better filtering is loadCustomerById(sale.customerId).
                // But we want to preserve sale's "snapshot" name if customer changed?
                // Actually `Sale` stores `customerName` at time of sale.
                // If we edit, we usually want current customer data? Or historical?
                // Usually we want to link to current customer.
                
                loadCustomerById(sale.customerId)

                // Async issue: loadCustomerById is async. fetchProductsForItems is async.
                // That's fine.
                
                fetchProductsForItems(items, 0, ArrayList())
            }

            override fun onFailure(e: Exception) {
                _isLoading.postValue(false)
                _error.postValue("Failed to load sale: " + e.message)
            }
        })
    }

    private fun loadCustomers() {
        viewModelScope.launch {
             customerRepository.getCustomers()
                 .onSuccess {
                     _customers.value = it
                 }
                 .onFailure {
                     _error.value = "Failed to load customers: ${it.message}"
                 }
        }
    }

    private fun fetchProductsForItems(items: List<SaleItem>, index: Int, accumulated: MutableList<ProductSelection>) {
        if (index >= items.size) {
            _productSelections.postValue(accumulated) // Use postValue if callback bg
            // Recalculate subtotal needs to happen on main thread if observing immediate
            // But we can trigger it.
            // Actually, we can't call recalculateSubtotal() easily from BG thread because it reads .value
            // Recommendation: Dispatch to main.
            // Since this is legacy callback hell, I'll leave it as is assuming main thread callbacks for now, 
            // but if I could, I'd refactor SaleRepository to suspend.
            
            // Hack for recalculate:
             _productSelections.postValue(accumulated)
             // We need to wait for LD update?
             // Or just calc manually
             var sub = 0.0
             for (item in accumulated) {
                sub += FinancialCalculator.calculateLineItemTotal(
                    item.product.sellingPrice,
                    item.quantityInSale
                )
             }
             _subtotal.postValue(sub)
             
            _isLoading.postValue(false)
            return
        }

        val item = items[index]
        saleRepository.getProduct(item.productId, object : SaleRepository.ProductCallback {
            override fun onSuccess(product: Product) {
                // Use fresh product data but maintain sale quantity
                val sel = ProductSelection(product, item.quantity)
                // Override selling price with the price from the sale item to reflect
                // historical price
                product.sellingPrice = item.pricePerItem

                accumulated.add(sel)
                fetchProductsForItems(items, index + 1, accumulated)
            }

            override fun onFailure(e: Exception) {
                // If product deleted or not found, create placeholder
                val placeholder = Product()
                placeholder.documentId = item.productId
                placeholder.name = item.productName + " (Unavailable)"
                placeholder.sellingPrice = item.pricePerItem
                placeholder.quantity = 0

                val sel = ProductSelection(placeholder, item.quantity)
                accumulated.add(sel)
                fetchProductsForItems(items, index + 1, accumulated)
            }
        })
    }
}
